An AI assisted PR to help resolving the reported issue number 30 (#37)
grafana/catchpoint-prometheus-exporter: README.md

- `--port` or `CATCHPOINT_EXPORTER_PORT`: Sets the port on which the exporter will run (default: `9090`).
- `--webhook-path` or `CATCHPOINT_WEBHOOK_PATH`: Defines the path where the exporter will receive webhook data from Catchpoint (default: `/catchpoint-webhook`).
- `--verbose` or `CATCHPOINT_VERBOSE`: Enables verbose logging to provide more detailed output for debugging purposes (default: `false`).

> - `--stale-timeout` or `CATCHPOINT_STALE_TIMEOUT`: How long a test/node result is kept after its last webhook (default: `24h`; `0s` keeps them forever). See [Staleness](#staleness).

Reviewer: Worlds tiniest nit: I meant to comment on this before, why are you using a semicolon here?

> - `--max-body-bytes` or `CATCHPOINT_MAX_BODY_BYTES`: Largest webhook body accepted, in bytes (default: `1048576`).

Reviewer: Nit
```suggestion
- `--max-body-bytes` or `CATCHPOINT_MAX_BODY_BYTES`: Webhook response max size, in bytes (default: `1048576`).
```

- `CATCHPOINT_WEBHOOK_PATH`: Overrides the default webhook path.
- `CATCHPOINT_VERBOSE`: Set to `true` to enable verbose logging.
- `CATCHPOINT_STALE_TIMEOUT`: Overrides the staleness timeout, e.g. `90m`. Defaults to `24h`.

## Metrics

The exporter provides a range of metrics, reflecting various performance aspects captured by Catchpoint. A complete list of available metrics can be found in the file [/collector/testdata/all_metrics.prom](/collector/testdata/all_metrics.prom).

Catchpoint sends one webhook per test run **per node**, and the exporter keeps the most recent result for every test/node combination it has seen. Each result becomes its own set of series, identified by the `test_id` and `node_id` labels, so a scrape of `/metrics` returns every test from every node — not only whichever one reported last. See [/collector/testdata/multi_series_metrics.prom](/collector/testdata/multi_series_metrics.prom) for what two tests across two nodes look like.

Every test metric carries these labels: `test_id`, `node_id`, `node_name`, `test_name`, `client_id`, `asn`, `division_id`, `monitor_type_id`, `type_id`.

The exporter also reports on itself:

> - `catchpoint_up`: always `1` while the exporter is able to serve a scrape.

Reviewer: This is a breaking change. It should be possible to go to 0 on misconfiguration and bad credentials. An always-up metric is less useful than current, which would show much more cleanly on dashboards/alerts than the absence of up

Author: this is back to default now

- `catchpoint_webhook_requests_total`: webhook requests received.
- `catchpoint_webhook_errors_total`: webhook requests that could not be processed (wrong method, undecodable body, or a payload with no `TestId`).
- `catchpoint_tracked_series`: number of test/node combinations currently being exported.

> If `catchpoint_tracked_series` stays at `1` while you expect many tests, the webhook template is most likely not sending `TestId`/`NodeId` — check that your template matches [template.json](/template.json), since those two fields are what separate one test's results from another's. Payloads without a `TestId` are rejected with HTTP 400 and counted in `catchpoint_webhook_errors_total`.

Reviewer: Can this ever happen in reality? (staying at 1)

As you rightly point out, the lack of `TestId` would trigger a webhook error and tracked in a separate metric.

A missing `NodeID` would collapse the `seriesKey`'s `{testId, nodeId}` relation into a single series per test?

As a result you'd only ever run into the case of staying at exactly 1 when/if there's no `nodeId` and only *one* test in the users dataset.

Suggest we reword this to explain the behaviour from a different perspective, e.g.:
- Tracked series is at 0 while webhooks are climbing, you're likely not emitting `TestID`
- Tracked series match number of tests exactly, rather than number of tests * number of nodes, then you're not emitting `nodeID` properly and overwriting each node?

Author: hopefully the rewording is ok now?

> ## Staleness

Reviewer: This is a great addition in general, thank you!

The exporter keeps the last result of every test/node it has seen and re-exports it on every scrape. Metric timestamps are always scrape time, not the time of the Catchpoint test run, so a result that arrived hours ago is indistinguishable downstream from one that arrived seconds ago — nothing in the exposed data reveals that a test has stopped reporting.

`--stale-timeout` bounds that. A test/node result stops being exported once this long has passed since its last webhook, so a test you delete in Catchpoint disappears from `/metrics` rather than freezing at its final value forever. The default is `24h`: long enough that a run of missed test executions never drops a live series, short enough that deleted tests retire within a day.

Set it lower if you want deleted or broken tests to disappear sooner — a few multiples of your slowest test frequency is the rule of thumb, so `--stale-timeout=90m` suits tests running every 15–30 minutes. **Never set it below your slowest test frequency**, or live series will flap in and out between runs.

> `--stale-timeout=0s` disables eviction entirely and restores the pre-`24h` behaviour of keeping every result until the process restarts. The exporter logs a warning at startup when you do. Note that eviction happens during a scrape, so an exporter that receives webhooks while nothing scrapes `/metrics` accumulates results regardless of this setting.

Reviewer: Thinking about it some more, I'd argue we should probably leave it disabled by default and mention stale timeout as an optional configuration so there's no breaking change slipping in when/if users update the exporter

Author: would you like to go back to 0 by default then? I think it is sensible to have as 24h timeout as Maria suggested?

Reviewer: Yes please, it's a potential breaking change that could surprise people if they don't notice it.

Feel free to add a readme saying suggested timeout if enabled

> 

Reviewer: Same as other AI smells. Too verbose / LLM-like in the writing. Please reduce the volume and be a little more concise to match the rest of the descriptions.

Saying that, I think having the expanded explanation here is a fantastic addition, just needs to be pruned a bit

Author: hopefully this is more concise

Reviewer: Generally improved but commented on some remaining smells

The exporter also reports on itself:

> - `catchpoint_up`: `1` while the exporter is running; `0` if the last webhook body could not be decoded, until one decodes again.

Reviewer: Nit: AI smell, please remove overuse of semi colons and overly verbose wording.

As a follow-up to this, the last comment I had here was that we should be clear about using `0` for credential / misconfiguration issue as well, so as an example here's how I would write this

```suggestion
- `catchpoint_up`: `1` if exporter is running. `0` if misconfigured, has invalid credentials, or is encountering webhook errors
```

- `catchpoint_webhook_requests_total`: webhook requests received.
- `catchpoint_webhook_errors_total`: webhook requests that could not be processed (wrong method, undecodable body, a payload with no `TestId`, or the series limit being reached).
- `catchpoint_series_dropped_total`: results dropped after reaching the series limit. See [Series limit](#series-limit).
- `catchpoint_tracked_series`: number of test/node combinations currently being exported.

> Compare `catchpoint_tracked_series` against what you expect, using [template.json](/template.json) as the reference:

Reviewer: Minor nit since I'm looking closely here anyways:
```suggestion
Compare `catchpoint_tracked_series` against what you expect using [template.json](/template.json) as the reference:
```

- Stuck at `0` while `catchpoint_webhook_requests_total` climbs: the template is not sending `TestId`. Those payloads are rejected with HTTP 400 and counted in `catchpoint_webhook_errors_total`.
- Equal to your number of tests, rather than tests × nodes: the template is not sending `NodeId`, so every node of a test collapses onto one series and overwrites the others.

## Series limit

The exporter keeps one entry per test/node combination it has seen, so an unsupervised deployment grows with whatever Catchpoint sends it. `--max-series` bounds that.

> At the limit, combinations already tracked keep updating; only unseen ones are rejected, with HTTP 503 and a bump to `catchpoint_series_dropped_total`. Existing tests keep reporting rather than the exporter going blind all at once.

Reviewer: this is an awkward wording. Consider rephrasing this paragraph entirely

> The default of `10000` is far above any realistic Catchpoint account, so reaching it usually means something is wrong — a template sending unexpected `TestId` values, or `--stale-timeout=0s` letting retired tests accumulate. A negative value disables the limit and logs a warning at startup.

Reviewer: Nit: more AI smells, please write your documentation yourself or at least rephrase it after having it fleshed out :) 

Additionally suggest removing the last sentence as per my other feedback about negative values being invalid configuration input
```suggestion
The default of `10000` series is far above any realistic Catchpoint account, so reaching it usually means something is wrong. 
This could be a template sending unexpected `TestId` values, or `--stale-timeout=0s` letting stale tests accumulate. 
```

## Staleness

The exporter re-exports the last result of every test/node on each scrape, timestamped at scrape time rather than test-run time. Nothing in the exposed data shows that a test has stopped reporting, so without eviction a deleted test keeps exporting its final value forever.

> `--stale-timeout` drops a result once that long has passed since its last webhook. Set it to a few multiples of your slowest test frequency — `90m` suits tests running every 15–30 minutes. Below that frequency, live series flap between runs.

Reviewer: Again, please remove AI smells like em-dashes and semi-colons
